### ISSUE TITLE
Make use of prepared commands

### ISSUE THREAD

**Issue by roji** [aspnet/EntityFramework#5459]
ADO.NET allows users to prepare commands, which allows databases to cache query plans and parses. IIRC in modern version of SqlServer this doesn't have much effect (any?), but in other providers this is an important feature - repeated complicated queries in PostgreSQL definitely benefit from prepared commands.

I know there's already an SQL caching mechanism inside EFCore, and although I'm sure it would be non-trivial to cache commands as well it would be a great feature.

Definitely a lower-priority backlog feature request.

**Comment by bricelam** [comment]
If we do this, we should definitely do aspnet/Microsoft.Data.Sqlite#235 :wink:

**Comment by roji** [comment]
FYI, I'm working on some benchmarking and performance aspects of Npgsql, and prepared command use in EFCore would potentially be huge. I hope it gets prioritized at some point.

**Comment by roji** [comment]
Note that for the next release of Npgsql (3.2), I plan to introduce an automatic/implicit preparation mechanism at the driver level - any statement used more than X times would be automatically prepared up to a certain quota, etc. This should increase performance with EFCore in a very significant way, without any change on your end.

However, it seems like EFCore may have more knowledge as to which statements to prepare - in any case it seems better-placed to make such decisions than the ADO.NET driver. So I think this feature is still relevant at the EFCore level.

**Comment by bricelam** [comment]
Microsoft.Data.Sqlite will support prepared statements in version 2.1.0. PR: aspnet/Microsoft.Data.Sqlite#388

**Comment by bricelam** [comment]
The biggest impact this feature would have on SQLite is in the update pipeline. Ironically, not batching statements, but instead re-using single-statement commands yields the best performance.

**Comment by roji** [comment]
@bricelam is it not possible to both batch and use prepared statements at the same time? With Npgsql/PostgreSQL this yields the best performance.

**Comment by bricelam** [comment]
SQLite doesn't actually support batching. 😀 We fake it in the ADO.NET layer.

**Comment by smitpatel** [comment]
Take it till you make it :trollface:

**Comment by roji** [comment]
Oh right, for a moment I forgot we're talking about a local database where batching would be if limited utility anyway :)

**Comment by P-DHrestak** [comment]
So many years have passed, is this still not supported?
After many attempts to find a way for Npgsql prepared statements to work with EF Core still no success.

Adding the automatic preparation as suggested in https://www.npgsql.org/doc/prepare.html inside the UseNpgsql connection string gave no performance benefits leading me to believe that EF Core is blocking this in some way.

Sad news :(

**Comment by hrestakd** [comment]
So many years have passed, is this still not supported?
After many attempts to find a way for Npgsql prepared statements to work with EF Core still no success.

Adding the automatic preparation as suggested in https://www.npgsql.org/doc/prepare.html inside the UseNpgsql connection string gave no performance benefits leading me to believe that EF Core is blocking this in some way.

Sad news :(

**Comment by roji** [comment]
@hrestakd if Npgsql's automatic preparation doesn't help, then non-automatic prepared statements most probably wouldn't help either. Either you're not executing the same SQLs enough times, so they don't get prepared (but then preparing them explicitly would be a bad idea too), or you have some performance bottleneck elsewhere that's preventing the preparation-related improvement from showing up. In any case, it's very unlikely that EF is blocking better perf for you.

I recommend carefully understand what's going on and why Npgsql's automatic preparation doesn't help.

**Comment by hrestakd** [comment]
@roji you are right, it was not EF Core issue at all. Apologies for dirtying this thread.
The performance suddenly jumped when I removed the Auto Prepare Min Usages=1 from the connection string. Not sure how that impacted me, but our queries jumped significantly.
